File: solution/java/CheckoutServiceFixed.java
package com.shopsphere.checkout.service;

import com.shopsphere.checkout.client.DiscountClient;
import com.shopsphere.checkout.client.PaymentClient;
import com.shopsphere.checkout.model.CheckoutResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

/**
 * SOLUTION: Fixed CheckoutService — Java
 * All 5 bugs resolved.
 */
@Service
public class CheckoutServiceFixed {

    private static final Logger logger = LoggerFactory.getLogger(CheckoutServiceFixed.class);

    @Autowired
    private PaymentClient paymentClient;

    @Autowired
    private DiscountClient discountClient;

    public CheckoutResponse processCheckout(int userId, double amount) {
        logger.info("Processing checkout for userId={} amount={}", userId, amount);

        try {
            // Step 1: Get discount
            Double discount = discountClient.getDiscount(userId);

            // FIX 1: Handle null discount — treat as 0.0 (no discount)
            double safeDiscount = (discount != null) ? discount : 0.0;
            double finalAmount = amount - safeDiscount;

            logger.info("Discount={} finalAmount={} for userId={}", safeDiscount, finalAmount, userId);

            // FIX 2: Await the async payment result with timeout
            CompletableFuture<String> paymentFuture = paymentClient.processPaymentAsync(userId, finalAmount);
            String paymentStatus = paymentFuture.get(5, TimeUnit.SECONDS);  // FIX 4: enforce timeout

            if (!"success".equals(paymentStatus)) {
                logger.error("Payment failed for userId={} status={}", userId, paymentStatus);
                return CheckoutResponse.builder()
                        .status("failed")
                        .userId(userId)
                        .build();
            }

            return CheckoutResponse.builder()
                    .status("success")
                    .userId(userId)
                    .amount(finalAmount)
                    .discountApplied(safeDiscount)
                    .transactionId("TXN-" + userId + "-" + System.currentTimeMillis())
                    .build();

        } catch (Exception e) {
            // FIX 3: Log exception with full context and stack trace
            logger.error("Checkout failed for userId={} amount={}", userId, amount, e);
            return CheckoutResponse.builder()
                    .status("failed")
                    .userId(userId)
                    .build();
        }
    }
}

File: solution/python/checkout_service_fixed.py
"""
SOLUTION: Fixed CheckoutService — Python
All 5 bugs resolved.
"""
import asyncio
import logging

from app.client.discount_client import DiscountClient
from app.client.payment_client import PaymentClient

logger = logging.getLogger("checkout_service")


class CheckoutService:
    def __init__(self):
        self.payment_client = PaymentClient()
        self.discount_client = DiscountClient()

    async def process_checkout(self, user_id: int, amount: float) -> dict:
        logger.info(f"Processing checkout for user_id={user_id} amount={amount}")

        try:
            # Step 1: Apply discount
            discount = self.discount_client.get_discount(user_id)

            # FIX 1: Handle None discount — treat as 0 (no discount)
            discount = discount if discount is not None else 0.0

            final_amount = amount - discount
            logger.info(f"Discount applied: {discount}. Final amount: {final_amount}")

            # FIX 2: Await the payment task — propagate failures to caller
            payment_result = await self.payment_client.process_payment(user_id, final_amount)

            if payment_result.get("status") != "success":
                logger.error(
                    f"Payment failed for user_id={user_id} amount={final_amount}: {payment_result}"
                )
                return {"status": "failed", "reason": "payment_declined"}

            return {
                "status": "success",
                "user_id": user_id,
                "amount": final_amount,
                "discount_applied": discount,
                "transaction_id": payment_result.get("transaction_id"),
            }

        except Exception as e:
            # FIX 3: Log exception with full context and stack trace
            logger.error(
                f"Checkout failed for user_id={user_id} amount={amount}: {e}",
                exc_info=True,
            )
            return {"status": "failed", "reason": str(e)}
